tests/v2/validation/charts/cisbenchmark_test.go

```diff
@@ -0,0 +1,198 @@
+	initialCisbenchmarkChart, err := extencharts.GetChartStatus(client, c.project.ClusterID, charts.CISBenchmarkNamespace, charts.CISBenchmarkName)
+	require.NoError(c.T(), err)
+
+	if initialCisbenchmarkChart.IsAlreadyInstalled {
+		c.T().Skip("Skipping fresh installtion. cis-benchmark chart is already installed.")
+	}
+
+	c.T().Log("Installing latest version of cis benchmark chart")
+	err = charts.InstallCISBenchmarkChart(client, c.chartInstallOptions)
+	require.NoError(c.T(), err)
+
+	c.T().Log("Waiting for cisbenchmark chart deployments to have expected number of available replicas")
+	err = extencharts.WatchAndWaitDeployments(client, c.project.ClusterID, charts.CISBenchmarkNamespace, metav1.ListOptions{})
+	require.NoError(c.T(), err)
+
+	c.T().Log("Running a scan")
+
+	// Run the scan
+	err = cis.RunCISScan(client, c.project.ClusterID, c.cisProfileName, true)
+	require.NoError(c.T(), err)
+
+}
+
+func (c *CisBenchmarkTestSuite) TestUpgradeCisbenchmarkChart() {
+	subSession := c.session.NewSession()
+	defer subSession.Cleanup()
+
+	client, err := c.client.WithSession(subSession)
+	require.NoError(c.T(), err)
+
+	// Change cisbenchmark install option version to previous version of the latest version
+	versionsList, err := client.Catalog.GetListChartVersions(charts.CISBenchmarkName, catalog.RancherChartRepo)
+	require.NoError(c.T(), err)
+
+	if len(versionsList) < 2 {
+		c.T().Skip("Skipping the upgrade case, only one version of cis-benchmark is available")
+	}
+
+	versionLatest := versionsList[0]
+	c.T().Log(versionLatest)
+
+	versionBeforeLatest := versionsList[1]
+	c.T().Log(versionBeforeLatest)
+	c.chartInstallOptions.Version = versionBeforeLatest
+
+	c.T().Log("Checking if the cis-benchmark chart is installed with one of the previous versions")
+	initialCisbenchmarkChart, err := extencharts.GetChartStatus(client, c.project.ClusterID, charts.CISBenchmarkNamespace, charts.CISBenchmarkName)
+	require.NoError(c.T(), err)
+
+	if initialCisbenchmarkChart.IsAlreadyInstalled && initialCisbenchmarkChart.ChartDetails.Spec.Chart.Metadata.Version == versionLatest {
+		c.T().Skip("Skipping the upgrade case, cis-benchmark chart is already installed with the latest version")
+	}
+
+	if !initialCisbenchmarkChart.IsAlreadyInstalled {
+		c.T().Log("Installing cis-benchmark chart with the version before the latest version")
+		err = charts.InstallCISBenchmarkChart(client, c.chartInstallOptions)
+		require.NoError(c.T(), err)
+
+		c.T().Log("Waiting cis-benchmark chart deployments to have expected number of available replicas")
+		err = extencharts.WatchAndWaitDeployments(client, c.project.ClusterID, charts.CISBenchmarkNamespace, metav1.ListOptions{})
+		require.NoError(c.T(), err)
+
+		c.T().Log("Waiting cisbenchmark chart DaemonSets to have expected number of available nodes")
+		err = extencharts.WatchAndWaitDaemonSets(client, c.project.ClusterID, charts.CISBenchmarkNamespace, metav1.ListOptions{})
+		require.NoError(c.T(), err)
+	}
```

Comment:
Can you have 2 tests for the upgrade use case:
* if chart is already installed, this will be a `TestDynamicUpgradeCisbenchmarkChart`
* if chart not installed, install the chart version n-1, then upgrade to n

Reply:
Are you suggesting writing two seperate test function for upgrade scenarios?
1. TestUpgradeCisBenchmarkChart - This will skip if any version of chart is installed. If the chart is not installed this function will install n-1 version of chart and upgrade it to latest.
2. TestDynamicUpgradeCisbenchmarkChart - This will upgrade the chart to latest version if any other version is installed.

Both will be skipped if latest version of chart is already installed.

Reply:
correct